### PR TITLE
Add throw point for `UnhandledMatchError`

### PR DESCRIPTION
**Pull request by rajyan**
fixes https://github.com/phpstan/phpstan/issues/6115

### CODE REVIEW

File: src/Analyser/NodeScopeResolver.php [hunk 1]
```diff
@@ -2572,6 +2575,9 @@ static function (Node $node, Scope $scope) use ($nodeCallback): void {
 				$throwPoints = array_merge($throwPoints, $armResult->getThrowPoints());
 				$matchScope = $matchScope->filterByFalseyValue($filteringExpr);
 			}
+			if (!$hasDefaultCond) {
+				$throwPoints[] = ThrowPoint::createExplicit($scope, new ObjectType(UnhandledMatchError::class), $expr, false);
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php:2579]: This isn't sufficient. It's possible that the match handles all the cases and doesn't throw the exception, even without the default cause.

See the rule on how I figure that out: https://github.com/phpstan/phpstan-src/blob/d0ab72ac815f4a8716fa110d5c1e6caba6ceeb47/src/Rules/Comparison/MatchExpressionRule.php#L82-L91

Additionally, I'd like the linked error to not be reported if the match is surrounded by a try-catch for `UnhandledMatchError`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/NodeScopeResolver.php (hunk 1 above)
Comment (rajyan) [reply src/Analyser/NodeScopeResolver.php:2579]: Thanks! I wasn't aware of this rule. I'll try to fix it sometime this week.

**Comment by rajyan** [comment]
> Additionally, I'd like the linked error to not be reported if the match is surrounded by a try-catch for UnhandledMatchError.

Traversed parents to find try-catch, but not sure this is a good way. Also, maybe this error should be suppressed if there is an explicit `@throws` tag with `UnhandledMatchError`?

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you! The `@throws` tag would be a nice addition in a next PR. You can check `$scope->getFunction()->getThrowType()` for that.

**Pull request merged**